### PR TITLE
add PDC_WIDE

### PR DESCRIPTION
**Pull request by laelnasan**
It should be enough to make pdcurses support utf8, but lacks testing

* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

### CODE REVIEW

File: src/core/pdcurses/curses.h [hunk 1]
```diff
@@ -35,6 +35,7 @@ PDCurses portable platform definitions list:
 #define SYSVcurses      1      /* System V Curses routines */
 #define BSDcurses       1      /* BSD Curses routines */
 #define CHTYPE_LONG     1      /* size of chtype; long */
+#define PDC_WIDE        1      /* wide-character support */
```
Comment (waruqi) [inline src/core/pdcurses/curses.h:38]: You should add it in xmake.lua

```
    if is_plat("windows") then
        add_defines("PDCURSES", "PDC_WIDE=1")
        add_includedirs("src/core/pdcurses")
    else
        add_packages("ncurses")
    end
```

But I tested it and it seems to have no effect.

Comment (laelnasan) [reply src/core/pdcurses/curses.h:38]: oh, that is sad... but I will investigate it further

**Comment by waruqi** [comment]
I have tested it, I works on win10, but dose not work on win7.

In addition, the text display is not complete on win10.

### win7

Snip20201207_29

### win10

Snip20201207_30

**Comment by laelnasan** [comment]
It seems like the win7 is not rendering properly, should be the lacking of fonts? There is one emoji printed properly, hence the console does receive the text properly...

**Comment by laelnasan** [comment]
have you tried https://stackoverflow.com/questions/388490/how-to-use-unicode-characters-in-windows-command-line/388500#388500 ?

**Comment by laelnasan** [comment]
I'm having a hard time trying to set up the dev environment in win7, perhaps you could land me a hand?

**Comment by waruqi** [comment]
> I'm having a hard time trying to set up the dev environment in win7, perhaps you could land me a hand?

Ok, will do.

**Comment by laelnasan** [comment]
> Ok, will do.

I'm trying to set up xmake in win7, but only documentation I found is for PowerShell. I tried to follow it but no success

**Comment by waruqi** [comment]
> have you tried https://stackoverflow.com/questions/388490/how-to-use-unicode-characters-in-windows-command-line/388500#388500 ?

I tried `chcp 65001`

Snip20201207_31

**Comment by waruqi** [comment]
> I'm trying to set up xmake in win7, but only documentation I found is for PowerShell. I tried to follow it but no success

You can download xmake installder to install it from https://github.com/xmake-io/xmake/releases

xmake-v2.3.9.win64.exe

**Comment by waruqi** [comment]
Is it because I did not install the Chinese language pack, which caused some fonts to be missing?

**Comment by laelnasan** [comment]
> You can download xmake installder to install it from https://github.com/xmake-io/xmake/releases
> 
> xmake-v2.3.9.win64.exe

Thanks!

> Is it because I did not install the Chinese language pack, which caused some fonts to be missing?

it does seems like there are fonts missing...

**Comment by waruqi** [comment]
> it does seems like there are fonts missing...

ok, I think as long as the corresponding font is installed, it should be able to work. We can focus on testing and improving support on win10 first.

**Comment by laelnasan** [comment]
> ok, I think as long as the corresponding font is installed, it should be able to work. We can focus on testing and improving support on win10 first.

Did the behaviour change at all?

**Comment by waruqi** [comment]
> Did the behaviour change at all? I'm concerned that the Lua byte for utf8 types are < 0 on windows, that would mess with `wcwidth()`

You can set `LTUI_LOGFILE=c:\log.txt` env and use `log:print` to show utf8 byte on windows.

**Comment by laelnasan** [comment]
unfortunately it seems like pdcurses doesn't render wide characters properly https://github.com/wmcbrine/PDCurses/issues/87 it can be poorly solved by printing extra spaces in front of fullwidth characters:
image

but I encountered some further issues with input that I have no idea of how to solve. I've found a modified version of pdcurses that seems to solve those issues https://github.com/Bill-Gray/PDCursesMod . I don't have the time to look too much into it right now, though.

**Comment by waruqi** [comment]
Ok, thanks!

**Pull request closed without merging**